src/jit/optimizer.cpp

```diff
@@ -5835,18 +5842,19 @@ bool Compiler::optNarrowTree(GenTree* tree, var_types srct, var_types dstt, Valu
                     {
                         dstt = genSignedType(dstt);
 
-                        if (oprSize == dstSize)
+                        if ((oprSize == dstSize) && (!varTypeIsUnsigned(dstt) || varTypeIsUnsigned(oprt) ||
```

Comment:
It's not clear to me why the signedness of `oprt` is relevant here. It's basically always signed (`TYP_INT` or `TYP_LONG`). Did you intend to test the `GTF_UNSIGNED` flag on the cast node perhaps? But that does not apply to cast to small types so it does not seem relevant.

Reply:
Hmm
* `!varTypeIsUnsigned(dstt)` should always be `true` due to the above `dstt = genSignedType(dstt);`
* `dstSize == genTypeSize(genActualType(dstt))` is really `!varTypeIsSmall(dstt)`

So the interesting case is `(oprSize == dstSize) && varTypeIsSigned(oprt)` when the current useless cast is transformed into a cast to `dstt` while the original cast (that triggered narrowing) is still getting removed.

So this does seem to be correct, at least in the case of `cast(cast(ind(x))`.

Reply:
>> !varTypeIsUnsigned(dstt) should always be true due to the above dstt = genSignedType(dstt);

genSignedType doesn't change the sign of small types, e.g., genSignedType(ushort) returns ushort.

Reply:
>> It's not clear to me why the signedness of oprt is relevant here. It's basically always signed (TYP_INT or TYP_LONG). 

No, that's not correct. It can be, e.g., TYP_SHORT or TYP_USHORT.

Reply:
For example, here
```
CAST      int <- ushort <- long
    CAST      long <- int
       expr short
```
oprt is TYP_SHORT

Reply:
>> dstSize == genTypeSize(genActualType(dstt)) is really !varTypeIsSmall(dstt)
Yes, I'll update that.

Reply:
> genSignedType doesn't change the sign of small types, e.g., genSignedType(ushort) returns ushort.

Eh, completely forgot about that.

> No, that's not correct. It can be, e.g., TYP_SHORT or TYP_USHORT.

I know but it's kind of confusing because the signedness of the cast operand doesn't determine the kind of conversion being performed so it doesn't seem fit with the comment below that says "Same size and there is no signed --> unsigned conversion for small types".

I don't know, I need to stare more at this, I feel I'm still missing something.

Reply:
In this example
```
CAST      int <- ushort <- long
    CAST      long <- int
       expr short
```
dsst is ushort
oprt is short

transforming it to expr short is incorrect because there is an implicit short -> int conversion.
So it needs to be
```
CAST      int <- ushort <- int
    expr short
```
On the other hand, for this example

```
CAST      int <- ushort <- long
    CAST      long <- int
       expr ushort
```
it's ok to transform it to
expr ushort.

Reply:
I understand the goal but I don't quite understand how we get there. As far as I can tell the morphed cast is still going to be removed and to compensate for that the useless `long<-int` cast is being changed to a cast to `dstt`. This is obvious correct for the given examples but it's not clear to me what happens for
* `CAST.short(CAST.long(IND.ushort))` - `dstt` is `short` so it is signed and the condition is true. The cast to long gets changed to a nop and the original cast is removed. We end up with just `IND.ushort`.
* What happens if we have AND/OR/XOR along the path between casts? It seems to me that the cast to `ushort` from your first example could move bellow something like `OR x, 0x10000` and as a result bit 16 set by OR will remain set because there's no longer a cast to clear it.

Reply:
The only 'expr short/ushort' that we shodul see are memory reads.

Reply:
I think in the second example the optimization will be blocked because when optNarrowTree is called on AND/OR/XОР it'll call optNarrowTree on 0x10000 with dsst ==TYP_USHORT and that will return false here: https://github.com/dotnet/coreclr/blob/10729e1a8a65bf3669049bca234187cbaf1a2e70/src/jit/optimizer.cpp#L5665

Reply:
> I think in the second example the optimization will be blocked because when optNarrowTree is called on AND/OR/XОR it'll call optNarrowTree on 0x10000 with dsst ==TYP_USHORT and that will return false here: 

Right, I've missed the fact that `optNarrowTree` does not attempt to narrow constants below `TYP_INT`.